Yosemite/YosemiteTests/SnapshotsProvider/FetchResultSnapshotsProviderTests.swift

```diff
@@ -0,0 +1,239 @@
+        cancellables.removeAll()
+        super.tearDown()
+    }
+
+    func test_objectWithID_returns_the_expected_immutable_object() throws {
+        // Given
+        let insertedAccount = insertAccount(displayName: "Reina Feil", username: "reinafeil")
+
+        viewStorage.saveIfNeeded()
+
+        let query = FetchResultSnapshotsProvider<StorageAccount>.Query(
+            sortDescriptor: .init(keyPath: \StorageAccount.displayName, ascending: true)
+        )
+        let provider = FetchResultSnapshotsProvider(storageManager: storageManager, query: query)
+
+        // When
+        let snapshot: FetchResultSnapshot = try waitFor { promise in
+            provider.snapshot.dropFirst().sink { snapshot in
+                promise(snapshot)
+            }.store(in: &self.cancellables)
+
+            try provider.start()
+        }
+
+        // Then
+        let objectID = try XCTUnwrap(snapshot.itemIdentifiers.first)
+        let fetchedAccount = try XCTUnwrap(provider.object(withID: objectID))
+        XCTAssertEqual(fetchedAccount, insertedAccount.toReadOnly())
+    }
+
+    func test_snapshot_emits_an_empty_list_if_SnapshotsProvider_is_not_started() throws {
+        // Given
+        insertAccount(displayName: "Reina Feil", username: "reinafeil")
+
+        let query = FetchResultSnapshotsProvider<StorageAccount>.Query(
+            sortDescriptor: .init(keyPath: \StorageAccount.displayName, ascending: true)
+        )
+        let provider = FetchResultSnapshotsProvider(storageManager: storageManager, query: query)
+
+        // When
+        let snapshot: FetchResultSnapshot = try waitFor { promise in
+            provider.snapshot.first().sink { snapshot in
+                promise(snapshot)
+            }.store(in: &self.cancellables)
+        }
+
+        // Then
+        assertEmpty(snapshot.itemIdentifiers)
+    }
+
+    func test_snapshot_can_emit_a_sorted_list() throws {
+        // Given
+        let accounts = [
+            insertAccount(displayName: "Reina Feil", username: "reinafeil"),
+            insertAccount(displayName: "Arvid Lowe", username: "arvidlowe"),
+            insertAccount(displayName: "Lee Johns", username: "leejohns")
+        ]
+
+        let query = FetchResultSnapshotsProvider<StorageAccount>.Query(
+            sortDescriptor: .init(keyPath: \StorageAccount.displayName, ascending: true)
+        )
+        let provider = FetchResultSnapshotsProvider(storageManager: storageManager, query: query)
+
+        // When
+        let snapshot: FetchResultSnapshot = try waitFor { promise in
+            provider.snapshot.dropFirst().sink { snapshot in
+                promise(snapshot)
+            }.store(in: &self.cancellables)
+
+            try provider.start()
+        }
+
+        // Then
+        let expectedObjectIDs = [accounts[1].objectID, accounts[2].objectID, accounts[0].objectID]
+        let actualObjectIDs = snapshot.itemIdentifiers
+        XCTAssertEqual(expectedObjectIDs, actualObjectIDs)
```

Comment:
super nit: I think we usually put the actual value before the expected value in `XCTAssertEqual`? Just thought it might be more consistent when reading the error message 🙂 (I think other frameworks/languages might use the reverse order? maybe the code base hasn't been using the more common convention? 😅 )

Reply:
Yea what a shame the official documentation doesn't have an example 🙄 I think it's muscle memory for me to always pass the actual value first, and I'll be really confused if I ever write in a different language that follows the opposite 😆